src/Mod/Sketcher/App/SketchObject.cpp

```diff
@@ -2327,125 +2328,128 @@ int SketchObject::delConstraintOnPoint(int VertexId, bool onlyCoincident)
+            return false;
+        }
+        // Assuming `constr` already involves geoId and posId, all conditions are already met
+        constr->substituteIndexAndPos(geoId, posId, replaceGeoId, replacePosId);
+        return true;
+    };
+
+    auto performAllConstraintChecksOrChanges = [&](auto& constr) -> std::optional<bool> {
+        if (constr->Type != Sketcher::Coincident && onlyCoincident) {
+            return true;
+        }
+        switch (constr->Type) {
+            case Sketcher::Coincident:
+                return performCoincidenceChecksOrChanges(constr);
+            case Sketcher::Distance:
+            case Sketcher::DistanceX:
+            case Sketcher::DistanceY: {
+                return (transferToReplacement(constr->First, constr->FirstPos)
+                        || transferToReplacement(constr->Second, constr->SecondPos));
+            }
+            case Sketcher::PointOnObject: {
+                return transferToReplacement(constr->First, constr->FirstPos);
+            }
+            case Sketcher::Tangent:
+            case Sketcher::Perpendicular: {
+                // we could keep this constraint by converting it to a simple one, but that doesn't
+                // always work (for example if tangent-via-point is necessary), and it is not really
+                // worth it
+                return false;
             }
-            else if ((*it)->Type == Sketcher::Symmetric) {
-                if (((*it)->First == GeoId && (*it)->FirstPos == PosId)
-                    || ((*it)->Second == GeoId && (*it)->SecondPos == PosId)) {
-                    continue;// skip this constraint
-                }
+            case Sketcher::Vertical:
+            case Sketcher::Horizontal:
+            case Sketcher::Symmetric: {
+                return false;
             }
-            else if ((*it)->Type == Sketcher::Vertical || (*it)->Type == Sketcher::Horizontal) {
-                if (((*it)->First == GeoId && (*it)->FirstPos == PosId)
-                    || ((*it)->Second == GeoId && (*it)->SecondPos == PosId)) {
-                    continue;// skip this constraint
-                }
+            default:
+                return std::nullopt;
+        }
+    };
+
+    // remove or redirect any constraints associated with the given point
+    for (auto& constr : vals) {
+        // keep the constraint if it doesn't involve the point
+        if (!constr->involvesGeoIdAndPosId(geoId, posId)) {
+            // for these constraints remove the constraint even if it is not directly associated
+            // with the given point
+            if ((constr->Type == Sketcher::Distance || constr->Type == Sketcher::DistanceX
+                 || constr->Type == Sketcher::DistanceY)
+                && (constr->First == geoId && constr->FirstPos == PointPos::none
+                    && (posId == PointPos::start || posId == PointPos::end))) {
+                continue;
             }

```

Comment:
This logic block is bit difficult to read. Splitting it up into named segments would help a lot:
 ```suggestion
            // for these constraints remove the constraint even if it is not directly associated
            // with the given point
            const bool isDistanceType = constr->Type == Sketcher::Distance
                || constr->Type == Sketcher::DistanceX
                || constr->Type == Sketcher::DistanceY;
            const bool involvesGeoWithoutPos =
                constr->First == geoId && constr->FirstPos == PointPos::none;
            const bool isPosAnEndpoint = posId == PointPos::start || posId == PointPos::end;
            if (isDistanceType && involvesGeoWithoutPos && isPosAnEndpoint) {
                continue;
            }
```

Reply:
Done locally.